File: src/main/java/com/philhanna/flashcards/session/TrackedCard.java
package com.philhanna.flashcards.session;

import com.philhanna.flashcards.*;
import com.philhanna.flashcards.deck.BasicCard;

/**
 * An implementation of the <code>SessionCard</code> interface that
 * combines the card functions (get question, get answer, toggle) with
 * the statistics-keeping required for a session.
 */
public class TrackedCard extends BasicCard implements SessionCard {

   // ==========================================================
   // Instance variables
   // ==========================================================

   private CardHistory lastAnswer = CardHistory.NEVER_ANSWERED;
   private int timesAnsweredRight = 0;
   private int timesAnsweredWrong = 0;

   // ==========================================================
   // Inner class that implements the CardStatistics interface
   // ==========================================================

   private CardStatistics statistics = new CardStatistics() {
      public CardHistory getHistory() {
         return TrackedCard.this.lastAnswer;
      }

      public int getTimesAnsweredRight() {
         return TrackedCard.this.timesAnsweredRight;
      }

      public int getTimesAnsweredWrong() {
         return TrackedCard.this.timesAnsweredWrong;
      }
   };

   // ==========================================================
   // Constructors
   // ==========================================================

   /**
    * Creates a new default session card implementation, delegating the
    * implementation of the card interface methods to a card object
    * passed as a parameter.
    * @param card the card it contains
    */
   public TrackedCard(Card card) {
      super(card.getQuestion(), card.getAnswer());
   }

   // ==========================================================
   // Instance methods
   // ==========================================================

   /**
    * Returns the card statistics
    */
   public CardStatistics getStatistics() {
      return this.statistics;
   }

   /**
    * Marks an answer as correct
    */
   public void markRight() {
      this.lastAnswer = CardHistory.ANSWERED_RIGHT;
      this.timesAnsweredRight += 1;
   }

   /**
    * Marks an answer as incorrect
    */
   public void markWrong() {
      this.lastAnswer = CardHistory.ANSWERED_WRONG;
      this.timesAnsweredWrong += 1;
   }
}

File: src/flashcards/domain/session/tracked_card.py
# flashcards.domain.session.tracked_card

from flashcards.domain.card import BasicCard, Card
from flashcards.domain.card_history import CardHistory
from flashcards.domain.card_statistics import CardStatistics
from flashcards.domain.session.session_card import SessionCard


class TrackedCard(BasicCard, SessionCard):

    def __init__(self, card: Card) -> None:
        super().__init__(card.question, card.answer)
        self._statistics = CardStatistics()

    @property
    def statistics(self) -> CardStatistics:
        return self._statistics

    def mark_right(self) -> None:
        self._statistics.history = CardHistory.ANSWERED_RIGHT
        self._statistics.times_answered_right += 1

    def mark_wrong(self) -> None:
        self._statistics.history = CardHistory.ANSWERED_WRONG
        self._statistics.times_answered_wrong += 1
